pkg/model/iam/iam_builder.go

```diff
@@ -292,6 +273,57 @@ func addRoute53ListHostedZonesPermission(p *IAMPolicy) {
 	})
 }
 
+func addS3Permissions(p *IAMPolicy, iamPrefix string, s3Path *vfs.S3Path, role api.InstanceGroupRole) {
+	// Note that the config store may itself be a subdirectory of a bucket
+	iamS3Path := s3Path.Bucket() + "/" + s3Path.Key()
+	iamS3Path = strings.TrimSuffix(iamS3Path, "/")
+
+	p.Statement = append(p.Statement, &IAMStatement{
+		Effect: IAMStatementEffectAllow,
+		Action: stringorslice.Of("s3:GetBucketLocation", "s3:ListBucket"),
+		Resource: stringorslice.Slice([]string{
+			strings.Join([]string{iamPrefix, ":s3:::", s3Path.Bucket()}, ""),
+		}),
+	})
+
+	p.Statement = append(p.Statement, &IAMStatement{
+		Effect: IAMStatementEffectAllow,
+		Action: stringorslice.Slice([]string{"s3:List*"}),
+		Resource: stringorslice.Of(
+			strings.Join([]string{iamPrefix, ":s3:::", iamS3Path}, ""),
+			strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/*"}, ""),
+		),
+	})
+
+	if role == api.InstanceGroupRoleMaster {
+		p.Statement = append(p.Statement, &IAMStatement{
+			Effect: IAMStatementEffectAllow,
+			Action: stringorslice.Slice([]string{"s3:Get*"}),
+			Resource: stringorslice.Of(
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/*"}, ""),
+			),
+		})
+	} else if role == api.InstanceGroupRoleNode {
+		p.Statement = append(p.Statement, &IAMStatement{
+			Effect: IAMStatementEffectAllow,
+			Action: stringorslice.Slice([]string{"s3:Get*"}),
+			Resource: stringorslice.Of(
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/addons/*"}, ""),
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/instancegroup/*"}, ""),
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/pki/issued/*"}, ""),
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/pki/ssh/*"}, ""),
+				strings.Join([]string{iamPrefix, ":s3:::", iamS3Path, "/pki/private/ca/*"}, ""),
```

Comment:
The `kops-configuration` service fails to start when it doesn't have access to the ca key. I'll dig through the repo and see if I can find out why it needs it.